### ISSUE TITLE
Failed: PersistentClusterShardingLeaving

### ISSUE THREAD

**Issue by chbatey** [akka/akka#24709]
https://jenkins.akka.io:8498/job/akka-release-2.4/479/consoleFull

```
[JVM-1] - must recover after leaving coordinator node *** FAILED ***
[JVM-1]   java.lang.RuntimeException: barrier failed: after-4
[JVM-1]   at akka.remote.testconductor.ClientFSM$$anonfun$3.applyOrElse(Player.scala:210)
[JVM-1]   at akka.remote.testconductor.ClientFSM$$anonfun$3.applyOrElse(Player.scala:185)
[JVM-1]   at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36)
[JVM-1]   at akka.actor.FSM$class.processEvent(FSM.scala:663)
[JVM-1]   at akka.remote.testconductor.ClientFSM.akka$actor$LoggingFSM$$super$processEvent(Player.scala:145)
[JVM-1]   at akka.actor.LoggingFSM$class.processEvent(FSM.scala:799)
[JVM-1]   at akka.remote.testconductor.ClientFSM.processEvent(Player.scala:145)
[JVM-1]   at akka.actor.FSM$class.akka$actor$FSM$$processMsg(FSM.scala:657)
[JVM-1]   at akka.actor.FSM$$anonfun$receive$1.applyOrElse(FSM.scala:651)
[JVM-1]   at akka.actor.Actor$class.aroundReceive(Actor.scala:502)
```

**Comment by raboof** [comment]
https://jenkins.akka.io:8498/job/akka-release-2.4/496/consoleFull

**Comment by raboof** [comment]
https://jenkins.akka.io:8498/job/akka-multi-node/3834/consoleFull

**Comment by jrudolph** [comment]
https://jenkins.akka.io:8498/job/akka-multi-node/3855/consoleFull

**Comment by jrudolph** [comment]
The original error is probably this one:

```
[info] [JVM-3] - must recover after leaving coordinator node (on node 'third', class akka.cluster.sharding.PersistentClusterShardingLeavingMultiJvmNode3) *** FAILED ***
[info] [JVM-3]   java.lang.AssertionError: assertion failed: timeout (1 second) during expectMsgClass waiting for class akka.actor.ActorRef
[info] [JVM-3]   at scala.Predef$.assert(Predef.scala:223)
[info] [JVM-3]   at akka.testkit.TestKitBase.expectMsgClass_internal(TestKit.scala:528)
[info] [JVM-3]   at akka.testkit.TestKitBase.expectMsgType(TestKit.scala:510)
[info] [JVM-3]   at akka.testkit.TestKitBase.expectMsgType$(TestKit.scala:509)
[info] [JVM-3]   at akka.testkit.TestKit.expectMsgType(TestKit.scala:928)
[info] [JVM-3]   at akka.cluster.sharding.ClusterShardingLeavingSpec.$anonfun$new$15(ClusterShardingLeavingSpec.scala:222)
[info] [JVM-3]   at scala.collection.immutable.HashMap$HashMap1.foreach(HashMap.scala:234)
[info] [JVM-3]   at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:468)
[info] [JVM-3]   at akka.cluster.sharding.ClusterShardingLeavingSpec.$anonfun$new$14(ClusterShardingLeavingSpec.scala:218)
[info] [JVM-3]   at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
```

**Comment by patriknw** [comment]
https://jenkins.akka.io:8498/job/akka-classic-cluster-tests/370

**Comment by johanandren** [comment]
https://jenkins.akka.io:8498/job/akka-nightly-jdk8/7809/consoleText

**Comment by johanandren** [comment]
https://jenkins.akka.io:8498/job/akka-artery-cluster-tests/2404/consoleText

**Comment by johanandren** [comment]
Looks like the two last is unrelated, and is about the region sending the `GracefulShutdownReq` to the cordinator when it is just right then starting up and waiting for the state from the replicator - the `GracefulShutdownReq` goes to deadletters and there is no retrying. Perhaps something we broke with the remember entities changes.

**Comment by johanandren** [comment]
Actually, it seems it is because of the remote flushing deathwatch #28695 fix

**Comment by johanandren** [comment]
Looks like @chbatey saw the same before, though not happen on each test run: https://github.com/akka/akka/issues/29212#issuecomment-644686019

**Comment by jrudolph** [comment]
https://jenkins.akka.io:8498/job/akka-nightly-2.13/648/consoleFull

**Comment by jrudolph** [comment]
https://jenkins.akka.io:8498/job/akka-multi-node/4196/consoleFull

**Comment by patriknw** [comment]
https://jenkins.akka.io:8498/job/akka-nightly-jdk8/8022

**Comment by johanandren** [comment]
Looking at the last logs, the two oldest nodes `JVM-[1,2]` are leaving (but not done with that), causing the coordinator to move to `JVM-3`, the leaving nodes then register with the new coordinator, which allocates shards to them that they have to start (and then gracefully stop, I think).

Part of this is solved by #29553 where the `LeastShardAllocationStrategy` does not consider those node for rebalance except as the last possible way out. But I wonder if there isn't something more here, a region will accept `ShardHome` even though `gracefulShutdownProgress` is true, shouldn't it refuse to host a shard in that scenario? (ping @patriknw, @renatocaval )

**Comment by ignasi35** [comment]
https://jenkins.akka.io:8498/job/akka-nightly-2.13/663/display/redirect
